src/main/java/agh/project/Map/AbstractWorldMap.java

```diff
@@ -0,0 +1,352 @@
+    private Map<Vector2d, Integer> deadPosition = new HashMap<>();
+
+    @Override
+    public boolean canMoveTo(Vector2d position) {
+        return (position.x >= 0 && position.x <= this.width && position.y >= 0 && position.y <= this.height);
+    }
+
+    @Override
+    public boolean place(Animal animal) {
+        return false;
+    }
+
+    @Override
+    public boolean isOccupiedByPlant(Vector2d position) {
+        return this.plants.containsKey(position);
+    }
+
+    @Override
+    public IMapElement objectAt(Vector2d position) {
+        if (this.isOccupiedByAnimal(position)) {
+            List<Animal> animalsList = animals.get(position);
+            return animalsList.get(0);
+        } else if (this.isOccupiedByPlant(position)) {
+            return plants.get(position);
+        } else {
+            return null;
+        }
+    }
+
+    @Override
+    public void specialMoves(Animal animal, Vector2d checkVector) {
+
+    }
+
+    @Override
+    public boolean isOccupiedByAnimal(Vector2d position) {
+        return this.animals.containsKey(position);
+    }
+
+    @Override
+    public boolean isOccupiedByAnimals(Vector2d position) {
+        return animals.get(position).size() > 1;
+    }
+
+    @Override
+    public List<Animal> animalsAt(Vector2d position) {
+        if (this.isOccupiedByAnimal(position)) {
+            return this.animals.get(position);
+        }
+        return null;
+    }
+
+    @Override
+    public List<Animal> copulation() {
+        List<Animal> newAnimals = new ArrayList<>();
+
+
+        for (List<Animal> animalList : animals.values()) {
+            boolean CanCopulate = true;
+
+            if (animalList.size() > 1) {
+                List<Animal> temporary = new ArrayList<>();
+                for (Animal animal : animalList) {
+
+                    if (animal.getEnergy() > energyToCopulate) {//wartość wczytana z pliku (energia konieczna do rozmnażania)
+                        temporary.add(animal);
+                    }
+                }
+                Animal animal1 = null;
+                Animal animal2 = null;
+                if (temporary.size() > 2) {
+                    animal1 = this.priority(temporary);
+                    temporary.remove(animal1);
+                    animal2 = this.priority(temporary);
+                } else if (temporary.size() == 2) {
+                    animal1 = temporary.get(0);
+                    animal2 = temporary.get(1);
+                } else {
+                    CanCopulate = false;
+                }
+                if (CanCopulate) {
+                    Animal newAnimal = animal1.copulation(animal2);
+                    if (newAnimal != null) {
+                        newAnimals.add(newAnimal);
+                    }
+                }
+            }
+        }
+        return newAnimals;
+    }
+
+    @Override
+    public Animal priority(List<Animal> animalsList) {
```

Comment:
to można by znacznie uprościć korzystając z tego nowego API do komparatorów z `.thenComparing(..)`.